Add New Heart and Valves Blocks based on the Regazzoni Heart Model

rjrios915 opened the pull request:
Add New Heart and Valves Blocks based on the Regazzoni Heart Model

## Current situation
I developed I model based on the equations outline in Regazzoni's heart model which required the implementation of a new heart block as well as new values.^1

1. Regazzoni, F., et al. “A cardiac electromechanical model coupled with a lumped-parameter model for closed-loop blood circulation.” Journal of Computational Physics, vol. 457, May 2022, p. 111083, https://doi.org/10.1016/j.jcp.2022.111083.


## Release Notes 
* Adds a elastic heart block 
* Adds a piecewise valve

## Documentation
Added Doxygen for all blocks.


## Testing
Added a closed loop test case validated against an existing model.


## Code of Conduct & Contributing Guidelines 
- [x] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

RickyAwesomeMan committed:
Added Reg CRL, Vessel, and Chambers

RickyAwesomeMan committed:
Added Reg CRL, Vessel, and Chambers

RickyAwesomeMan committed:
added CRL test cases

RickyAwesomeMan committed:
added CRL test cases

rjrios915 committed:
added a test case for CRL blood vessel block

rjrios915 committed:
added a test case for CRL blood vessel block

rjrios915 committed:
Added option to specify cardiac period

rjrios915 committed:
Added option to specify cardiac period

rjrios915 committed:
updated code

rjrios915 committed:
updated code

rjrios915 committed:
Fixes clang error on cardiac period #126

rjrios915 committed:
Fixes clang error on cardiac period #126

rjrios915 committed:
Fixes second clang error #126

rjrios915 committed:
Fixes second clang error #126

rjrios915 committed:
Colon switched for semi-colon #126

rjrios915 committed:
Colon switched for semi-colon #126

rjrios915 committed:
Merge branch 'SimVascular:master' into RegChamber

rjrios915 committed:
Removed cardiac period

rjrios915 committed:
Removed cardiac period

rjrios915 committed:
Merge branch 'RegChamber' of https://github.com/rjrios915/svZeroDSolver into RegChamber

rjrios915 committed:
merging RegChamber

rjrios915 committed:
Updated CRLVessel Documentation

rjrios915 committed:
Updated CRLVessel Documentation

ncdorn commented:
@rjrios915 can you rebase this branch and confirm all tests are passing?

rjrios915 committed:
Merge branch 'SimVascular:master' into RegChamber

rjrios915 committed:
Updated Reg Test Case

rjrios915 committed:
Updated Reg Test Case

rjrios915 committed:
clang format

rjrios915 committed:
clang format

rjrios915 committed:
Manually Revise ClangFormat

rjrios915 committed:
Manually Revise ClangFormat

rjrios915 committed:
updates

rjrios915 committed:
updates

rjrios915 committed:
Merge branch 'master' into RegChamber

ncdorn commented:
@mrp089 this has been rebased are you able to give the final review?

mrp089 reviewed (changes requested):
Thank you for your work, @rjrios915 and @ncdorn! I have a few (major and minor) ideas below that shouldn't be too complicated to change and will make the code much cleaner.

**Major**
- I highly appreciate Francesco's work. However, I wouldn't name the chamber and valve after someone, given that both are simple mathematical expressions that likely have been used by many researchers, and they're not the point of the paper cited. You should still cite the paper, but give the features a more descriptive name that makes it easier to understand what they're doing (e.g., step function valve).
- From looking at the equations, it seems to me that `RegazzoniValve` is very similar to `ValveTanh` and `RegazzoniChamber` to `ChamberElastanceInductance`. I thus propose to implement a class structure similar to `ClosedLoopCoronaryBC`: Have base classes `Valve` and `ChamberElastance` with different variants inheriting from them. These would also be named the same in the input files with the features that distinguish them (e.g., valve smooth or step) as an input parameter.

**Minor**
- I couldn't immediately see how `BloodVesselCRL` is different from the existing `BloodVessel`. If there are differences, please put that in a separate PR.
- Change all license headers to the new short format (#178).

ncdorn commented:
@mrp089 thanks for your feedback! 
- With regards to the base class `Valve` and `Chamber` with different variants - on Tuesday @alexkaiser and I were discussing something along these lines to address #157 so I will speak with Ricky and we will get that rolling. I think this is a nice and efficient solution to the current issue of having many different valve and chamber variants in the literature.
- We did find that the arrangement of the RCL circuit affected the results, and so we implemented `BloodVesselCRL` to verify our results against the literature. It is clunky but I wasn't able to see another way around unless we add some `arrangement` parameter to `BloodVessel`. We will open a separate PR for this!

ncdorn commented:
@rjrios915 and I were discussing today and one issue we (mostly Ricky) have ran into while implementing the heart chamber inheritance blocks is that if we prescribe different types of valve functions or elastance functions, these will potentially bring about a different set of parameters. Additionally I do not think we can provide strings as values in the input json, thus requiring the user to provide some integer or bool to the block in order to specify the elastance or valve function type. These ints inducing different function types, as well as their associated unique set of parameters, could create very confusing documentation and convoluted code readability. This could be remedied by the use of unique blocks and associated documentation for the different valve and chamber types. I think  the difficulty in reconciling the differences outweighs the inherent similarities between these various valve and chamber types and we should just preserve the current system of unique block types for each valve and chamber model. @mrp089 let me know your thoughts on this

rjrios915 committed:
Added option to specify cardiac period (#127)

Co-authored-by: ricky <rickyjrios@gmail.com>
Co-authored-by: Nick Dorn <ndorn@stanford.edu>

mrp089 commented:
Makes sense! Let's go ahead with the other minor changes and get this PR merged.

ncdorn committed:
Add GitHub workflows back and update testing (#192)

rjrios915 committed:
Renamed to piecewise and removed CRL from this branch

rjrios915 committed:
Updated Doc for Chamber

rjrios915 committed:
Clang

rjrios915 committed:
clang

rjrios915 committed:
Valve documentation

ncdorn committed:
Merge branch 'RegChamber' of https://github.com/rjrios915/svZeroDSolver into RickyRegChamber

rjrios915 committed:
clang and doc

rjrios915 committed:
Documentation updates

rjrios915 committed:
Doc

aabrown100-git commented:
@rjrios915 @ncdorn why was this PR closed with merging?

ncdorn commented:
> @rjrios915 @ncdorn why was this PR closed with merging?

I think @rjrios915 renamed some stuff, ricky can you open a new PR with the Piecewise Chamber Valve naming so we can merge that in?

rjrios915 commented:
@aabrown100-git @ncdorn Yes! I updated the naming and split the vessel and chamber/valve into different branches so ill create two new requests now that the window tests are updated!
